quic/s2n-quic-platform/src/syscall.rs

```diff
@@ -0,0 +1,176 @@
+            let socket = Socket::new(domain, socket_type, protocol)?;
+        } else {
+            let socket = Socket::new(domain, socket_type, protocol)?;
+            socket.set_nonblocking(true)?;
+        }
+    };
+
+    // allow ipv4 to also connect
+    if addr.is_ipv6() {
+        socket.set_only_v6(false)?;
+    }
+
+    socket.set_reuse_address(true)?;
+
+    #[cfg(unix)]
+    socket.set_reuse_port(reuse_port)?;
+
+    // mark the variable as "used" regardless of platform support
+    let _ = reuse_port;
+
+    socket.bind(&addr.into())?;
+
+    Ok(socket)
+}
+
+/// Disables MTU discovery and fragmentation on the socket
+pub fn configure_mtu_disc(tx_socket: &Socket) -> bool {
+    let mut success = false;
+
+    //= https://www.rfc-editor.org/rfc/rfc9000#section-14
+    //# UDP datagrams MUST NOT be fragmented at the IP layer.
+
+    //= https://www.rfc-editor.org/rfc/rfc9000#section-14
+    //# In IPv4 [IPv4], the Don't Fragment (DF) bit MUST be set if possible, to
+    //# prevent fragmentation on the path.
+
+    //= https://www.rfc-editor.org/rfc/rfc8899#section-3
+    //# In IPv4, a probe packet MUST be sent with the Don't
+    //# Fragment (DF) bit set in the IP header and without network layer
+    //# endpoint fragmentation.
+
+    //= https://www.rfc-editor.org/rfc/rfc8899#section-4.5
+    //# A PL implementing this specification MUST suspend network layer
+    //# processing of outgoing packets that enforces a PMTU
+    //# [RFC1191][RFC8201] for each flow utilizing DPLPMTUD and instead use
+    //# DPLPMTUD to control the size of packets that are sent by a flow.
+    #[cfg(s2n_quic_platform_mtu_disc)]
+    {
+        use std::os::unix::io::AsRawFd;
+
+        // IP_PMTUDISC_PROBE setting will set the DF (Don't Fragment) flag
+        // while also ignoring the Path MTU. This means packets will not
+        // be fragmented, and the EMSGSIZE error will not be returned for
+        // packets larger than the Path MTU according to the kernel.
+        success |= libc!(setsockopt(
+            tx_socket.as_raw_fd(),
+            libc::IPPROTO_IP,
+            libc::IP_MTU_DISCOVER,
+            &libc::IP_PMTUDISC_PROBE as *const _ as _,
+            core::mem::size_of_val(&libc::IP_PMTUDISC_PROBE) as _,
+        ))
+        .is_ok();
+
+        success |= libc!(setsockopt(
+            tx_socket.as_raw_fd(),
+            libc::IPPROTO_IPV6,
+            libc::IPV6_MTU_DISCOVER,
+            &libc::IP_PMTUDISC_PROBE as *const _ as _,
+            core::mem::size_of_val(&libc::IP_PMTUDISC_PROBE) as _,
+        ))
+        .is_ok();
+    }
+
+    success
+}
+
+/// Configures the socket to return TOS/ECN information as part of the ancillary data
+pub fn configure_tos(rx_socket: &Socket) -> bool {
+    let mut success = false;
+
+    #[cfg(s2n_quic_platform_tos)]
+    {
+        use std::os::unix::io::AsRawFd;
+        let enabled: libc::c_int = 1;
+
+        success |= libc!(setsockopt(
```

Comment:
this previously had the check for `if rx_addr.is_ipv4() || !cfg!(any(target_os = "macos", target_os = "ios"))`. Why is that removed?

Reply:
we're no longer bailing on errors; just returning if they succeeded or not. this new approach is more robust since we try our best to enable everything we can and fall back if it wasn't successful.